### PR TITLE
Running wp-e2e-tests in wp-calypso

### PR DESCRIPTION
**Pull request by bsessions85**
#### Changes proposed in this Pull Request

This PR pulls in all of the e2e tests from wp-e2e-tests. The CircleCI config has been updated to run Canary tests ( IE, Safari, and WooCommerce) as well as the full suite of e2e tests. 

More updates will come in future PRs. These include:
* Clean up documentation
* Clean up code so some eslint overrides can be removed
* Removing unneeded scripts

p6fDka-Bw-p2

#### Testing instructions

First, Ensure that tests run and pass in CircleCI

Test them locally:
* run `npm run test-e2e` and make sure tests start to run. Chrome instances should start opening and performing actions. 

Run a single spec file:
* run `cd test/e2e`
* run `./node_modules/.bin/mocha specs/wp-page-editor-spec.js`

Most of the documentation in the README with regards to running tests is still accurate as long as you are in the `test/e2e` directory

It isn't necessary that all of the tests run and pass locally as this can take a while and be resource intensive, but as long as they start and pass in CircleCI, it is good.

### CODE REVIEW

File: packages/i18n-calypso/package-lock.json [hunk 1]
```diff
@@ -180,9 +180,9 @@
 			},
```
Comment (simison) [inline packages/i18n-calypso/package-lock.json]: Changes to this file seem unrelated?

File: test/e2e/.babelrc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (simison) [inline test/e2e/.babelrc]: Splitting configs to multiple files is always unfortunate instead of having one file with visibility to all configs at once as we have with our eslint file.

via https://babeljs.io/docs/en/config-files#monorepos

>Separately, it is also important to decide if you want to use .babelrc files or just a central babel.config.js. .babelrc files are not required for subfolder-specific configuration like they were in Babel 6, so often they are not needed in Babel 7, in favor of a babel.config.js.

Could this be moved to root config under "overrides"?

File: test/e2e/.gitignore [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+.DS_Store
+temp/
+npm-debug.log
+
+# Ignore all local config files, but keep the example there.
+config/local-*
+
+# http://direnv.net
+.envrc
+/temp-cookies/
```
Comment (simison) [inline test/e2e/.gitignore]: We're prolly inheriting lots of this from root `.gitignore` already?

File: test/e2e/config/default.json [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		"privateByDefault",
+		"jetpackFreePlanButtonPosition",
+		"removeUsername",
+		"removeDomainsStepFromOnboarding",
+		"showConciergeSessionUpsell",
+		"showConciergeSessionUpsellNonGSuite",
+		"builderReferralThemesBanner",
+		"gSuitePlan",
+		"domainSearchButtonStyles"
+	],
```
Comment (sirreal) [inline test/e2e/config/default.json]: ✨ Oh hey, we wouldn't need to keep this in sync anymore:

https://github.com/Automattic/wp-calypso/blob/417ea8219d37955b5b7e842dc9ba7e7a73097c17/client/lib/abtest/active-tests.js#L3

(not a change request, just a bonus to pulling e2e in)

File: test/e2e/config/default.json [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	"overrideABTests": [
+		[ "businessPlanDescriptionAT_20170605", "original" ],
+		[ "skipThemesSelectionModal_20170904", "show" ],
```
Comment (sirreal) [inline test/e2e/config/default.json]: Same here! All this could go!

https://github.com/Automattic/wp-calypso/blob/417ea8219d37955b5b7e842dc9ba7e7a73097c17/client/lib/abtest/active-tests.js#L9

(not a change request, just a bonus to pulling e2e in)

File: test/e2e/docs/miscellaneous.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+- [List of wrapper repos & friends](#list-of-wrapper-repos--friends)
+- [What to name your branch](#what-to-name-your-branch)
+
+## NodeJS Version
+
+The node version should be defined in the `.nvmrc` file for use with the [nvm](https://github.com/creationix/nvm) project.
+
+## Git Pre-Commit Hook
+
+The file `/scripts/git-pre-commit-circleci-validate` will run `circleci validate` against the CircleCI config file prior to every commit.  This prevents the constant back-and-forth when making updates only to find that they fail immediately on CI.  Instructions in the file direct how to install the hook in your local Git environment (it won't run without this).
```
Comment (sirreal) [inline test/e2e/docs/miscellaneous.md]: We'll need to update this, I'd consider dropping the config validation on pre-commit. We don't check valid config in Calypso right now, most folks won't have the `circleci` available locally, and we don't want to slow down other commits.

File: .eslintrc.js [hunk 7]
```diff
@@ -62,6 +77,8 @@ module.exports = {
 		COMMIT_SHA: true,
 		// this is when Webpack last built the bundle
 		BUILD_TIMESTAMP: true,
+		// this is for e2e tests
+		step: false,
```
Comment (sirreal) [inline .eslintrc.js]: Can we set `globals` in the `overrides` for e2e above? If not, let's add an eslint config at the e2e root which contains all of the config.

I don't like the idea of adding a new global everywhere that isn't actually available and has a pretty generic name 🙂

**Review by sirreal — COMMENTED** [review]
I left some notes, this is really exciting to see. Can we get a nice description written of what's going on here?

Does this bring in all the core e2e tests in addition to all of the other wrapper repos like canary, ie11, branches…

File: packages/i18n-calypso/package-lock.json (hunk 1 above)
Comment (bsessions85) [reply packages/i18n-calypso/package-lock.json]: yes, thanks! removed it

File: test/e2e/.babelrc (hunk 2 above)
Comment (bsessions85) [reply test/e2e/.babelrc]: I tried this and wasn't able to get it to work. I think it is because the tests are still being run from the new subdirectory rather than the root, so the babel.config.js isn't checked. I can look in to this further for a future iteration

File: test/e2e/.gitignore (hunk 3 above)
Comment (bsessions85) [reply test/e2e/.gitignore]: Thanks, moved it all up to the root

File: .circleci/config.yml [hunk 8]
```diff
@@ -487,7 +523,19 @@ workflows:
       - prime-calypso-live:
           requires:
             - test-client
-            - test-server
+           # - test-server
```
[the remark this answers is not in the document]
Comment (bsessions85) [reply .circleci/config.yml]: Oops! This was part of debugging since that part kept failing and blocking me. Forgot to remove it when I got it passing. Thanks!

File: test/e2e/config/default.json (hunk 4 above)
Comment (bsessions85) [reply test/e2e/config/default.json]: Yeah, We will definitely make updates to use the source rather than relying on this config

File: .eslintrc.js (hunk 7 above)
Comment (bsessions85) [reply .eslintrc.js]: Yeah, good suggestion. Moved it to overrides and it works wonders. Thanks!

File: test/e2e/docs/miscellaneous.md (hunk 6 above)
Comment (bsessions85) [reply test/e2e/docs/miscellaneous.md]: Good call. I have removed the pre-commits, and the mention of it in docs

**Comment by Stojdza** [comment]
Tested locally and it works as expected. Great work @bsessions85!

Note: When I pulled branch and installed dependencies, I couldn't run `npm run test-e2e` because Magellan and Mocha were missing. I removed node_modules and installed dependencies again and it worked.

**Comment by Stojdza** [comment]
Tested locally after new changes and works well 👍 

Some CircleCI jobs run longer than expected. I've restarted `test-e2e-canary-woo` and it passed regularly.

File: .circleci/config.yml [hunk 9]
```diff
@@ -449,14 +460,43 @@ jobs:
...
-            curl --silent "https://hash-${CIRCLE_SHA1}.calypso.live" > /dev/null
-            exit 0
+          command: ./test/e2e/scripts/wait-for-running-branch.sh
+
+  test-e2e:
+      <<: *defaults
+      working_directory: ~/wp-calypso/test/e2e
+      parallelism: 2
+      steps:
+        - prepare
```
Comment (blowery) [inline .circleci/config.yml]: who in here is responsible for splitting tests between the two containers?

**Comment by blowery** [comment]
I think you can drop the change to calypso's `npm-shrinkwrap.json` entirely. We didn't add any new packages to package.json in the root, so the lock shouldn't change..

File: .circleci/config.yml (hunk 9 above)
Comment (bsessions85) [reply .circleci/config.yml]: We run desktop in one container and mobile in another

**Comment by bsessions85** [comment]
Removed the changes to npmshrinkwrap.json

File: .circleci/config.yml (hunk 9 above)
Comment (sirreal) [reply .circleci/config.yml]: Does that happen by checking the CircleCI node env variable somewhere in a script?

If there's such a clear split, should we make e2e a workflow that includes independent desktop and mobile jobs?

Comment (bsessions85) [reply .circleci/config.yml]: > Does that happen by checking the CircleCI node env variable somewhere in a script?

Yes, there is a script that checks which CircleCI node index it is running on and runs the desktop or mobile tests based on that value.

>If there's such a clear split, should we make e2e a workflow that includes independent desktop and mobile jobs?
That is certainly an option. I am open to whatever makes things more efficient and more clear for everyone.

Comment (sirreal) [reply .circleci/config.yml]: It seems to me like that would be clearer, but it's not something that needs to happen in this PR.

Comment (bsessions85) [reply .circleci/config.yml]: Yeah, I'll do that as a part of the task to add the triggers.

File: test/e2e/.babelrc (hunk 2 above)
Comment (jsnajdr) [reply test/e2e/.babelrc]: Babel looks for the `babel.config.js` config file only in the current working directory, so the file in `..` is not used. There's a `rootMode: 'upward'` option [to override that](https://babeljs.io/docs/en/config-files#root-babelconfigjs-files) in monorepos. Would it be useful for us?

Why do we use Babel at all? Is there any JSX or other non-standard syntax to transpile? Running on recent 10.x Node has the benefit of full support for latest standard JS without transpilation.

File: test/e2e/.babelrc [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+	"presets": [
+		["@babel/preset-env", {
+			"targets": {
+				"node": "10.6.0"
```
Comment (jsnajdr) [inline test/e2e/.babelrc]: We should use `node: 'current'` here. That means that Node version that's used to run Babel will be used as transpilation target. Doesn't need to be changed on every Node upgrade.

But as I mentioned above, maybe we don't need Babel at all in e2e?

Comment (bsessions85) [reply test/e2e/.babelrc]: We are really using babel for the import/export stuff. I am not very familiar with how it all works, but I think that is all we really need. I went ahead and set the rootMode and was able to remove the .babelrc file altogether. Thanks!

**Review by Stojdza — APPROVED** [review]
LGTM 🤞
